### PR TITLE
Optee smt unitary test

### PR DESCRIPTION
**Pull request by vingu-linaro**
Add unitary test for optee/smt module

The unitary test covers 100% line code

### CODE REVIEW

**Comment by vingu-linaro** [comment]
@mohamedasaker-arm 
This is the 3rd unitary test for optee/smt

File: module/optee/smt/test/mod_optee_smt_unit_test.c [hunk 1]
```diff
@@ -0,0 +1,889 @@
...
+    TEST_ASSERT_EQUAL(FWK_SUCCESS, status);
+}
+
+/*!
+ * \brief test module's bind fucntion for round 0.
+ *
+ *  \details Test bind for round 0 and others except for round 1 when it needs
+ *  additional information from bind_request of other modules.
+ */
+void test_mailbox_smt_bind_round0(void)
```
Comment (mohamedasaker-arm) [inline module/optee/smt/test/mod_optee_smt_unit_test.c:162]: The function name doesn't reflect this part. "and others except for round 1"
Can we say something like ..bind_rounds_except_round1?

Comment (mohamedasaker-arm) [reply module/optee/smt/test/mod_optee_smt_unit_test.c:162]: or maybe bind_trivial_rounds
anything indicate that it's not only round0 and they are of a specific nature

File: module/optee/smt/test/mod_optee_smt_unit_test.c [hunk 2]
```diff
@@ -0,0 +1,889 @@
...
+    device_config =
+        (struct mod_optee_smt_channel_config *)smt_element_table[FAKE_SCMI_SERVICE_IDX_OSPM_0]
+            .data;
+
+    /* Save policies */
+    saved_policies = device_config->policies;
+
+    /* Return secure state of the channel */
+    device_config->policies = MOD_SMT_POLICY_SECURE;
+    secure = 0xAA;
```
Comment (mohamedasaker-arm) [inline module/optee/smt/test/mod_optee_smt_unit_test.c:364]: why not true/false? and I believe false here would be better

File: module/optee/smt/test/mod_optee_smt_unit_test.c [hunk 3]
```diff
@@ -0,0 +1,889 @@
...
+    fwk_id_get_element_idx_ExpectAndReturn(channel_id, FAKE_SCMI_SERVICE_IDX_OSPM_0);
+    status = smt_get_secure(channel_id, &secure);
+    TEST_ASSERT_EQUAL(FWK_SUCCESS, status);
+    TEST_ASSERT_EQUAL(true, secure);
+
+    /* Restore policies */
+    device_config->policies = saved_policies;
+
+    /* Return non secure state of the channel */
+    secure = 0xAA;
```
Comment (mohamedasaker-arm) [inline module/optee/smt/test/mod_optee_smt_unit_test.c:375]: why not secure=true?

File: module/optee/smt/test/mod_optee_smt_unit_test.c [hunk 4]
```diff
@@ -0,0 +1,889 @@
...
+    /* Fail if the offset is out of range */
+    channel_id = FWK_ID_ELEMENT(FWK_MODULE_IDX_OPTEE_SMT, FAKE_SCMI_SERVICE_IDX_OSPM_0);
+    fwk_id_get_element_idx_ExpectAndReturn(channel_id, FAKE_SCMI_SERVICE_IDX_OSPM_0);
+    status = smt_write_payload(channel_id, channel_ctx->max_payload_size+1, fake_payload, 0);
+    TEST_ASSERT_EQUAL(FWK_E_PARAM, status);
+
+    /* Fail if the size is out of range */
+    channel_id = FWK_ID_ELEMENT(FWK_MODULE_IDX_OPTEE_SMT, FAKE_SCMI_SERVICE_IDX_OSPM_0);
+    fwk_id_get_element_idx_ExpectAndReturn(channel_id, FAKE_SCMI_SERVICE_IDX_OSPM_0);
+    status = smt_write_payload(channel_id, 0, (const void*)fake_payload, channel_ctx->max_payload_size+1);
```
Comment (mohamedasaker-arm) [inline module/optee/smt/test/mod_optee_smt_unit_test.c:546]: Minor: format:  `channel_ctx->max_payload_size + 1`

File: module/optee/smt/test/mod_optee_smt_unit_test.c [hunk 5]
```diff
@@ -0,0 +1,889 @@
...
+    fake_size = 0x0000;
+    fwk_id_get_element_idx_ExpectAndReturn(channel_id, FAKE_SCMI_SERVICE_IDX_OSPM_0);
+    status = smt_get_payload(channel_id, (const void **)&fake_payload, &fake_size);
+    TEST_ASSERT_EQUAL(FWK_SUCCESS, status);
+    TEST_ASSERT_EQUAL(((struct mod_optee_smt_memory *)(channel_in_buffer))->payload, fake_payload);
+    TEST_ASSERT_EQUAL(((struct mod_optee_smt_memory *)(channel_in_buffer))->length - sizeof(uint32_t), fake_size);
+
+}
+
+void fake_fwk_str_memcpy(void *dest, const void *src, size_t count, int iteration)
```
Comment (mohamedasaker-arm) [inline module/optee/smt/test/mod_optee_smt_unit_test.c:511]: Minor:
As I understand from your PR: you try to make helper/callback functions before usage
However, I don't see it in `test_smt_write_payload_early` So could you move it before `test_smt_write_payload`?

File: module/optee/smt/test/mod_optee_smt_unit_test.c [hunk 6]
```diff
@@ -0,0 +1,889 @@
...
+}
+
+void test_reset_out_buffer(void)
+{
+    int i;
+
+    for (i = 0; i < FAKE_MAILBOX_SIZE; i++) {
+        channel_out_buffer[i] = 0;
+    }
+}
```
Comment (mohamedasaker-arm) [inline module/optee/smt/test/mod_optee_smt_unit_test.c:646]: I suggest for these helper functions to remove the "test" prefix
to not conflict with the real tests.

**Review by mohamedasaker-arm — CHANGES_REQUESTED** [review]

**Comment by nicola-mazzucato-arm** [comment]
Agreed with @vingu-linaro to bring this PR over gitlab.
Thanks

**Comment by leandro-arm** [comment]
# Hello! 👋

Thank you for your contribution to the SCP-firmware Project!

SCP-firmware contributions can only be submitted through [GitLab](https://gitlab.arm.com/firmware/SCP-firmware). This GitHub repo is accessible as a Read-Only mirror and it is not able to accept contributions for the project.

This PR has been migrated to: https://gitlab.arm.com/firmware/SCP-firmware/-/merge_requests/834

The SCP-firmware team

**Pull request closed without merging**